inginious/frontend/plugins/task_hints/pages/user_hint_manager.py

```diff
@@ -0,0 +1,134 @@
+
+from inginious.frontend.parsable_text import ParsableText
+
+class UserHintManager(object):
+
+    """
+        Manage the user hints information in database, as the hints
+        that was unlocked, it's individual penalty, and cumulative
+        penalty, to be applied in the student's final grade for the
+        respective task.
+    """
+
+    def __init__(self, username, task_id, database):
+
+        self._username = username
+        self._task_id = task_id
+        self._database = database
+
+    def check_locked_hint_status(self, hints):
+
+        """ Method to check each hint status, and return the content
+            if it is allowed
+        """
+        user_hints = self._database.user_hints.find_one({"taskid":  self._task_id,
+                                            "username":  self._username})
+        if user_hints is None:
+            user_hints = self.create_hint_list_for_user()
+            return 200, ""
+
+        self.update_allowed_hints_list(user_hints, hints)
+
+        to_show_hints = {}
+        allowed_hints = user_hints["allowedHints"]
+        for key in hints:
+            to_show_hint_content = {
+                "title": hints[key]["title"],
+                "content": None,
+                "penalty": hints[key]["penalty"],
+                "allowed_to_see": False,
+            }
+            if hints[key]["id"] in [hint["id"] for i, hint in enumerate(allowed_hints)]:
+
+                parse_content = self.parse_rst_content(hints[key]["content"])
+                to_show_hint_content["content"] = parse_content
+                to_show_hint_content["allowed_to_see"] = True
+                to_show_hint_content["penalty"] = [hint["penalty"] for i, hint in enumerate(allowed_hints) if hint["id"] == hints[key]["id"]]
+
+            to_show_hints[key] = to_show_hint_content
+
+        return to_show_hints
+
+    def update_allowed_hints_list(self, user_hints, hints):
+
+        """ Method to compare the user hints with the task hints, and update them"""
+        allowedHints = user_hints["allowedHints"]
+        hints_id = {key:hint["id"] for key, hint in hints.items()}
```

Comment:
```suggestion
        hints_ids = {hint["id"] for hint in hints.values()}
``` 
That way you create a set of ids and you avoid the call to the values in the `for` structure as that is inefficient.